File: src/ElectronNET.Host/ElectronHostAPI/src/notification.ts
import { Notification } from 'electron';
import { Socket } from 'node:net';

const notifications: Electron.Notification[] = (global['notifications'] = global['notifications'] || []) as Electron.Notification[];
let electronSocket;

export function notificationApi(socket: Socket) {
    electronSocket = socket;
    socket.on('createNotification', (options) => {
        const notification = new Notification(options);
        let haveEvent = false;

        if (options.showID) {
            haveEvent = true;
            notification.on('show', () => {
                electronSocket.emit('NotificationEventShow', options.showID);
            });
        }

        if (options.clickID) {
            haveEvent = true;
            notification.on('click', () => {
                electronSocket.emit('NotificationEventClick', options.clickID);
            });
        }

        if (options.closeID) {
            haveEvent = true;
            notification.on('close', () => {
                electronSocket.emit('NotificationEventClose', options.closeID);
            });
        }

        if (options.replyID) {
            haveEvent = true;
            notification.on('reply', (event, value) => {
                electronSocket.emit('NotificationEventReply', [options.replyID, value]);
            });
        }

        if (options.actionID) {
            haveEvent = true;
            notification.on('action', (event, value) => {
                electronSocket.emit('NotificationEventAction', [options.actionID, value]);
            });
        }

        if (haveEvent) {
            notifications.push(notification);
        }

        notification.show();
    });

    socket.on('notificationIsSupported', () => {
        const isSupported = Notification.isSupported;
        electronSocket.emit('notificationIsSupportedComplete', isSupported);
    });
}

File: src/ElectronNET.API/API/Notification.cs
using ElectronNET.API.Entities;
using ElectronNET.API.Serialization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

namespace ElectronNET.API
{
    /// <summary>
    /// Create OS desktop notifications
    /// </summary>
    public sealed class Notification
    {
        private static Notification _notification;
        private static object _syncRoot = new object();

        internal Notification()
        {
        }

        internal static Notification Instance
        {
            get
            {
                if (_notification == null)
                {
                    lock (_syncRoot)
                    {
                        if (_notification == null)
                        {
                            _notification = new Notification();
                        }
                    }
                }

                return _notification;
            }
        }

        private static List<NotificationOptions> _notificationOptions = new List<NotificationOptions>();

        /// <summary>
        /// Create OS desktop notifications
        /// </summary>
        /// <param name="notificationOptions"></param>
        public void Show(NotificationOptions notificationOptions)
        {
            GenerateIDsForDefinedActions(notificationOptions);

            BridgeConnector.Socket.Emit("createNotification", notificationOptions);
        }

        private static void GenerateIDsForDefinedActions(NotificationOptions notificationOptions)
        {
            bool isActionDefined = false;

            if (notificationOptions.OnShow != null)
            {
                notificationOptions.ShowID = Guid.NewGuid().ToString();
                isActionDefined = true;

                BridgeConnector.Socket.Off("NotificationEventShow");
                BridgeConnector.Socket.On<string>("NotificationEventShow", (id) => { _notificationOptions.Single(x => x.ShowID == id).OnShow(); });
            }

            if (notificationOptions.OnClick != null)
            {
                notificationOptions.ClickID = Guid.NewGuid().ToString();
                isActionDefined = true;

                BridgeConnector.Socket.Off("NotificationEventClick");
                BridgeConnector.Socket.On<string>("NotificationEventClick", (id) => { _notificationOptions.Single(x => x.ClickID == id).OnClick(); });
            }

            if (notificationOptions.OnClose != null)
            {
                notificationOptions.CloseID = Guid.NewGuid().ToString();
                isActionDefined = true;

                BridgeConnector.Socket.Off("NotificationEventClose");
                BridgeConnector.Socket.On<string>("NotificationEventClose", (id) => { _notificationOptions.Single(x => x.CloseID == id).OnClose(); });
            }

            if (notificationOptions.OnReply != null)
            {
                notificationOptions.ReplyID = Guid.NewGuid().ToString();
                isActionDefined = true;

                BridgeConnector.Socket.Off("NotificationEventReply");
                BridgeConnector.Socket.On<string[]>("NotificationEventReply", (args) =>
                {
                    _notificationOptions.Single(x => x.ReplyID == args[0]).OnReply(args[1]);
                });
            }

            if (notificationOptions.OnAction != null)
            {
                notificationOptions.ActionID = Guid.NewGuid().ToString();
                isActionDefined = true;

                BridgeConnector.Socket.Off("NotificationEventAction");
                BridgeConnector.Socket.On<string[]>("NotificationEventAction", (args) =>
                {
                    _notificationOptions.Single(x => x.ActionID == args[0]).OnAction(args[1]);
                });
            }

            if (isActionDefined)
            {
                _notificationOptions.Add(notificationOptions);
            }
        }

        /// <summary>
        /// Whether or not desktop notifications are supported on the current system.
        /// </summary>
        /// <returns></returns>
        public Task<bool> IsSupportedAsync()
        {
            var tcs = new TaskCompletionSource<bool>();

            BridgeConnector.Socket.Once<bool>("notificationIsSupportedComplete", tcs.SetResult);
            BridgeConnector.Socket.Emit("notificationIsSupported");

            return tcs.Task;
        }


    }
}
